File: naver_news.py
# -*- coding: utf8 -*-

import requests
from bs4 import BeautifulSoup
import sys
import re
import time
from datetime import datetime,timedelta
import json,io

if len(sys.argv) !=4:
	print "usage : naver_news.py {start_date} {end_date} {query_text}"
	print "ex) naver.news.py 2016-10-01 2016-10-30 helloworld"
	exit()

query = sys.argv[3]

start_date = datetime.strptime(sys.argv[1],'%Y-%m-%d') - timedelta(days=1)
end_date = datetime.strptime(sys.argv[2],'%Y-%m-%d')

now = datetime.now()

lists=[]

while 1:
	start_date = start_date + timedelta(days=1)

	if start_date == end_date or start_date > now :
		break;
	
	date_str = start_date.strftime('%Y-%m-%d')

	baseurl = 'http://news.naver.com/main/search/search.nhn?query='+query+'&st=news.all&q_enc=EUC-KR&r_enc=UTF-8&r_format=xml&rp=none&sm=title.basic&ic=all&so=rel.dsc&detail=1&pd=4&r_cluster2_start=1&r_cluster2_display=10&start=1&display=10&startDate='+date_str+'&endDate='+date_str+'&dnaSo=rel.dsc&page='


	plain = requests.get(baseurl).text


	soup = BeautifulSoup(plain,'lxml')

	result_num = soup.find_all("span",class_="result_num")



	try: 
		# 1000건 이상시 1,000 으로 표현될때
		total_num = int(re.findall('\d,\d*',result_num[0].string)[0].replace(',',''))+1
	except:
		# 1000건 미만시 
		total_num = int(re.findall('\/\s*\d*',result_num[0].string)[0].split('/ ')[1])+1

	page_num = 1+ (total_num / 10)

	cnt=0
	for i in range(1,page_num+1):

		baseurl = 'http://news.naver.com/main/search/search.nhn?query='+query+'&st=news.all&q_enc=EUC-KR&r_enc=UTF-8&r_format=xml&rp=none&sm=title.basic&ic=all&so=rel.dsc&detail=1&pd=4&r_cluster2_start=1&r_cluster2_display=10&start=1&display=10&startDate='+date_str+'&endDate='+date_str+'&dnaSo=rel.dsc&page='+str(i)
		plain = requests.get(baseurl).text
		soup = BeautifulSoup(plain,'lxml')


		for list in soup.find_all("a",class_="tit"):
			tmp = {'link':list.get('href'),'title':list.string,'date':date_str}
			lists.append(tmp)
			#print tmp
			cnt+=1

		
		time.sleep(1)

	print 'total : '+str(cnt) + ' / date : '+date_str


with io.open('output.json','w',encoding='utf-8') as f:
	f.write(unicode(json.dumps(lists,ensure_ascii=False)))

File: naver_news.R

library(rvest)
library(ggplot2)
library(stringr)
library(jsonlite)
library(ggiraph)
library(dplyr)


start_dt = '2016-10-01'
end_dt = '2016-10-31'
query_str = 'naver'

days <- seq(from=as.Date(start_dt), to=as.Date(end_dt),by='days' )
date_v = c()
title_v = c()
for ( i in seq_along(days) )
{
  
  url = paste("http://news.naver.com/main/search/search.nhn?query=",query_str,"&st=news.all&q_enc=EUC-KR&r_enc=UTF-8&r_format=xml&rp=none&sm=title.basic&ic=all&so=rel.dsc&detail=1&pd=4&r_cluster2_start=1&r_cluster2_display=10&start=1&display=10&startDate=",days[i],"&endDate=",days[i],"&dnaSo=rel.dsc&page=",sep="")
 
  dt = read_html(url)
 
  result_num = dt %>% html_nodes(".result_num") %>% html_text()
 
  
  ret = tryCatch({
    result_num = strtoi(str_match(result_num,"\\/\\s*(\\d*)")[,2])  
  }, error = function(err){
    return("not found!")
    
  })
  
  if(ret =="not found!") next
  
  page_num = ceiling((result_num/10))
  
  cnt=0

  
  for( j in 1:page_num)
  {
    print(j)
    url = paste("http://news.naver.com/main/search/search.nhn?query=",query_str,"&st=news.all&q_enc=EUC-KR&r_enc=UTF-8&r_format=xml&rp=none&sm=title.basic&ic=all&so=rel.dsc&detail=1&pd=4&r_cluster2_start=1&r_cluster2_display=10&start=1&display=10&startDate=",days[i],"&endDate=",days[i],"&dnaSo=rel.dsc&page=",j,sep="")
    print(url)
    dt = read_html(url)
    
    lists = dt %>% html_nodes(".tit") %>% html_text()
    
    for (list in lists)
    {
      title_v = c(title_v,list) 
      date_v = c(date_v,as.character(days[i]))
     
    }
    
  }
  

}

result = data.frame(date_v,title_v)


# Year Split
result$date_v=str_split_fixed(result$date_v, "-",2)[,2]

# Date Count
b= result %>%
  group_by(date_v) %>%
  distinct(title_v) %>%
  summarise(merge=paste(title_v,collapse="<br>"),
            count=n())

b$merge=gsub("'","",b$merge)


g=ggplot(data=b,aes(x=date_v,y=count,group=1))+
  xlab("날짜") + ylab("건수") +
  geom_line(colour="red")+
  theme_bw()+
  geom_point_interactive(aes(tooltip=merge),size=0.5,alpha=0.5)+
  theme(axis.text.x = element_text(angle=70,hjust=1),text = element_text(size=8))

ggiraph(code=print(g),width=1,height=3)

